buildSrc/src/main/groovy/org/gradle/testing/IntegrationTest.groovy

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing
+
+import groovy.transform.CompileStatic
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+
+/**
+ * Verifies the correct behavior of a feature, as opposed to just a small unit of code.
+ * Usually referred to as 'functional tests' in literature, but our code base has historically
+ * been using the term 'integration test'.
+ */
+@CacheableTask
+@CompileStatic
+class IntegrationTest extends DistributionTest {
+    @Input
+    boolean requiresSamples
+
+    void setRequiresSamples(boolean requiresSamples) {
+        this.requiresSamples = requiresSamples
+        if (requiresSamples) {
+            dependsOn(':docs:userguideDocbook')
+        }
```

Comment:
This way I can't pass `false` to this once `true` has already been passed. I wonder if we can do something like this instead in the constructor:

```groovy
dependsOn { requiresSamples ? ":docs:userguideDocbook" : null }
```

It's also weird to encode the name of a specific task to depend on in the task type. What is the output we are using in this task from `:docs:userguideDocbook`? Can't we express this as a `FileCollection` passed as a property to this task, which would be `buildBy` the `:docs:userguideDocbook` task (or something else)?

It also seems weird why we'd be depending a task about Docbook to get our samples. Something about this needs clearing up.